### ISSUE TITLE
Gradient doesn't satisfy Taylor expansion

### ISSUE THREAD

**Issue by mloubout** [pytorch/pytorch#74688]
### 🐛 Describe the bug

I wanted to verify my code because I was obtaining some odd results so I decided to verify the gradient of a given layer is accurate. While pytorch has an internal check based on FD, my goto check for any gradient (ML or not) is to verify the linearization error. Is created a small script to verify the gradient for a simple singe Linear layer and unfortunately the gradient computed with `.backward` does not satisfy the linearization theory. See below the script verifying that the gradient satisfies the Taylor linearized expansion that fails for the gradient

```
import torch
import torch.nn as nn

n = (51, 51)

for dtype in [torch.float32, torch.float64]:
    print("%s, \n \n" % dtype)
    # Set intial weights and perturbation 
    w0 = 1 + torch.rand(*n, dtype=dtype)
    dw = 1e-2*(1 + torch.rand(*n, dtype=dtype))
    # Initial inputs
    x0 = 1 + torch.rand(*n, dtype=dtype)
    x1 = 1 + torch.rand(*n, dtype=dtype)
    # Create layer
    L = nn.Linear(*n, bias=False, dtype=dtype)

    # Loss function
    def loss(x, y):
        try:
            L.weight.grad.data.zero_()
        except AttributeError:
            pass
        return .5*torch.norm(L(x) - y)**2

    # Intialize layer
    L.weight.data = w0
    # Loss and gradient at x1
    y0 = L(x0)
    f0 = loss(x1, y0)
    f0.backward()
    g0 = L.weight.grad.detach()
    gdx = torch.dot(dw.flatten(), g0.flatten())
    # COmpute taylor error for varying step size
    h = 1
    # < g, dw>
    f0, gdx = float(f0.detach()), float(gdx.detach())
    # Loop of perturbation size
    nh = 7
    err1 = torch.zeros(nh, dtype=dtype)
    err2 = torch.zeros(nh, dtype=dtype)

    print("%8s %8s %8s %8s %8s %8s %8s" % ("h", "f0", "fk", "e1", "gdx", " h gdx", "e2"))
    for i in range(nh):
        L.weight.data = w0 + h*dw
        # Get f(x+dx))
        fk = loss(x1, y0)
        fk = float(fk.detach())
        print("%2.2e %2.2e %2.2e %2.2e %2.2e %2.2e %2.2e" % (h, f0, fk, fk-f0, gdx, h*gdx, fk - f0 - h*gdx))
        # First order error
        # Per taylor, we have `f(x+dx) = f(x) + O(h)` so this should decrease linearly with h
        err1[i] = abs(fk - f0)
        # Second order error
        # Per taylor, we have `f(x+dx) = f(x) + h*<g, dw> + O(h^2)` so this should decrease
        # quadratically with h
        err2[i] = abs(fk - f0 - h*gdx)
        h = h / 2

    # Check error rates. We should have err1 decrease wit h and err2 with h^2
    # So if we compute the ration we should get `2` and `4` since
    # `err1[i-1]/err1[i] = O(2*h)/ O(h) = 2`
    # `err2[i-1]/err2[i] = O((2*h)**2)/ O(h**2) = 4`
    print("%8s %8s %8s" % ("h", "p1", "p2"))
    h = .5
    for i in range(1, nh):
        print("%2.2e %2.2e %2.2e " % (h, float(err1[i-1]/err1[i]), float(err2[i-1]/err2[i])))
        h = h / 2


    ####################################################################################
    # We show here that if we do it by hand the gradient does satisfy it

    def lin_loss(A, x, y):
        return .5*torch.norm(x @ A.t() - y)**2

    # Use the exact same matrix and perturbation
    A = L.weight.data.detach()
    A.requires_grad = True

    y0 = x0 @ A.t()
    f0 = lin_loss(A, x1, y0)
    g0 = x1.t() @ (x1 @ A.t() - y0)

    gdx = torch.dot(dw.flatten(), g0.flatten())

    h = 1
    f0, gdx = float(f0.detach()), float(gdx.detach())

    err1 = torch.zeros(nh)
    err2 = torch.zeros(nh)

    print("%8s %8s %8s %8s %8s %8s %8s" % ("h", "f0", "fk", "e1", "gdx", " h gdx", "e2"))
    for i in range(nh):
        Aloc = A + h*dw
        # Get f(x+dx))
        fk = lin_loss(Aloc, x1, y0)
        fk = float(fk.detach())
        print("%2.2e %2.2e %2.2e %2.2e %2.2e %2.2e %2.2e" % (h, f0, fk, fk-f0, gdx, h*gdx, fk - f0 - h*gdx))
        # First order error
        # Per taylor, we have `f(x+dx) = f(x) + O(h)` so this should decrease linearly with h
        err1[i] = abs(fk - f0)
        # Second order error
        # Per taylor, we have `f(x+dx) = f(x) + h*<g, dw> + O(h^2)` so this should decrease
        # quadratically with h
        err2[i] = abs(fk - f0 - h*gdx)
        h = h / 2

    # Check error rates. We should have err1 decrease wit h and err2 with h^2
    # So if we compute the ration we should get `2` and `4` since
    # `err1[i-1]/err1[i] = O(2*h)/ O(h) = 2`
    # `err2[i-1]/err2[i] = O((2*h)**2)/ O(h**2) = 4`
    print("%8s %8s %8s" % ("h", "p1", "p2"))
    h = .5
    for i in range(1, nh):
        print("%2.2e %2.2e %2.2e " % (h, float(err1[i-1]/err1[i]), float(err2[i-1]/err2[i])))
        h = h / 2
```

For completeness , the exact same layer is implemented by hand with matrix vector products and shows that the decay rate f error should indeed satisfy the expansion. Where does that inaccuracy comes from?

### Versions

PyTorch version: 1.11.0
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 10.15.7 (x86_64)
GCC version: (Homebrew GCC 9.4.0) 9.4.0
Clang version: 12.0.0 (clang-1200.0.32.29)
CMake version: version 3.22.1
Libc version: N/A

Python version: 3.9.9 (main, Nov 21 2021, 03:22:47)  [Clang 12.0.0 (clang-1200.0.32.29)] (64-bit runtime)
Python platform: macOS-10.15.7-x86_64-i386-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.20.3
[pip3] torch==1.11.0
[conda] blas                      1.0                         mkl  
[conda] mkl                       2018.0.2                      1  
[conda] mkl-include               2019.0                      118  
[conda] mkl-service               1.1.2            py36h7ea6df4_4  
[conda] mkl_fft                   1.0.1            py36h917ab60_0  
[conda] mkl_random                1.0.1            py36h78cc56f_0  
[conda] numpy                     1.14.3           py36h9bb19eb_1  
[conda] numpy-base                1.14.3           py36h479e554_1  
[conda] numpydoc                  0.8.0                    py36_0  
[conda] pytorch                   0.4.1           py36_cuda0.0_cudnn0.0_1    pytorch
[conda] torchvision               0.2.1                    py36_1    pytorch
